### PR TITLE
network: TC introduce FIFO

### PR DESCRIPTION
**Pull request by ssahani**

### CODE REVIEW

**Comment by evverx** [comment]
I'm just going to close and reopen the PR to see whether it helps to trigger LGTM against 4ac78b1a37c5d3953f5d34aa40439dd1a68e536a.

**Comment by evverx** [comment]
Looks like it didn't work out. @ssahani could you force-push the code to make LGTM update its base commit?

**Comment by ssahani** [comment]
Sure thanks

**Comment by evverx** [comment]
Thank you! It seems to be working. Unfortunately, it means that It appears the only way to make LGTM "green" in https://github.com/systemd/systemd/pull/14854 and https://github.com/systemd/systemd/pull/14858 is to force-push the code there as well.

**Comment by ssahani** [comment]
> Thank you! It seems to be working. Unfortunately, it means that It appears the only way to make LGTM "green" in #14854 and #14858 is to force-push the code there as well.

Sure let me do that too.

File: man/systemd.network.xml [hunk 1]
```diff
@@ -2494,6 +2494,32 @@
     </variablelist>
   </refsect1>
 
+  <refsect1>
+    <title>[FIFO] Section Options</title>
+    <para>The <literal>[FIFO]</literal> section manages the queueing discipline (qdisc) of
+    First in First Out (FIFO).</para>
```
Comment (yuwata) [inline man/systemd.network.xml]: First In First Out

File: src/network/networkd-network-gperf.gperf [hunk 2]
```diff
@@ -303,6 +303,8 @@ TokenBucketFilter.LatencySec,                config_parse_token_bucket_filter_la
 TrivialLinkEqualizer.Parent,                 config_parse_qdisc_parent,                                QDISC_KIND_TEQL,               0
 TrivialLinkEqualizer.Handle,                 config_parse_qdisc_handle,                                QDISC_KIND_TEQL,               0
 TrivialLinkEqualizer.Id,                     config_parse_trivial_link_equalizer_id,                   QDISC_KIND_TEQL,               0
+FIFO.Parent,                                 config_parse_qdisc_parent,                                QDISC_KIND_FIFO,               0
```
Comment (yuwata) [inline src/network/networkd-network-gperf.gperf:306]: Please rebase and add `FIFO.Handle=`.

File: src/network/tc/fifo.c [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        if (r == -ENOMEM)
+                return log_oom();
+        if (r < 0)
+                return log_syntax(unit, LOG_ERR, filename, line, r,
+                                  "More than one kind of queueing discipline, ignoring assignment: %m");
+
+        fifo = FIFO(qdisc);
+
+        if (isempty(rvalue)) {
+                if (streq(lvalue, "PacketLimit"))
```
Comment (yuwata) [inline src/network/tc/fifo.c]: Please drop this, as this function is only used for `PacketLimit=`.

File: src/network/tc/fifo.c [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+        r = parse_size(rvalue, 1000, &k);
+        if (r < 0) {
+                log_syntax(unit, LOG_ERR, filename, line, r,
+                           "Failed to parse '%s=', ignoring assignment: %s",
+                           lvalue, rvalue);
+                return 0;
+        }
+
+        if (streq(lvalue, "PacketLimit"))
```
Comment (yuwata) [inline src/network/tc/fifo.c]: Same here.

File: src/network/tc/fifo.h [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+#pragma once
+
+#include "conf-parser.h"
+#include "qdisc.h"
+#include "time-util.h"
+
+typedef struct FirstInFirstOut {
+        QDisc meta;
+
+        size_t limit;
```
Comment (yuwata) [inline src/network/tc/fifo.h]: `uint32_t`

File: src/network/tc/fifo.c [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+                fifo->limit = k;
```
Comment (yuwata) [inline src/network/tc/fifo.c]: Before assigning value, please check the value range.

File: src/network/tc/fifo.c [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+        if (isempty(rvalue)) {
+                if (streq(lvalue, "PacketLimit"))
+                        fifo->limit = 0;
+
+                qdisc = NULL;
+                return 0;
+        }
+
+        r = parse_size(rvalue, 1000, &k);
```
Comment (yuwata) [inline src/network/tc/fifo.c]: Or maybe, `safe_atou32()`?

File: src/network/tc/fifo.c (hunk 6 above)
Comment (ssahani) [reply src/network/tc/fifo.c]: changed to safe_atou32.

**Comment by ssahani** [comment]
updated thanks for the review @yuwata .

**Comment by yuwata** [comment]
LGTM. As we already released rc1, let's merge this after v245. Is it OK for you?

**Comment by ssahani** [comment]
yes let's do after it

File: src/network/tc/fifo.c [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        FirstInFirstOut *fifo;
+        int r;
+
+        assert(link);
+        assert(qdisc);
+        assert(req);
+
+        fifo = FIFO(qdisc);
+
+        if (fifo->limit > 0)
```
Comment (yuwata) [inline src/network/tc/fifo.c:29]: This condition can be dropped. But OK.

**Comment by yuwata** [comment]
Included in #14890. Closing.

**Pull request closed without merging**